packages/components/src/components/Tabs/hooks/useTabTriggers.tsx

```diff
@@ -0,0 +1,139 @@
+        if (willFitAllItemsToTheLeft) {
+            activeIndicatorElement.style.left = `${activeTriggerElement.offsetLeft}px`;
+        } else {
+            activeIndicatorElement.style.left = `${triggerListElement.offsetWidth - activeTriggerElement.offsetWidth}px`;
+        }
+        activeIndicatorElement.style.width = `${activeTriggerElement?.offsetWidth}px`;
+    } else if (isOverflowingRight) {
+        activeIndicatorElement.style.left = `${triggerListElement.clientWidth - activeTriggerElement.offsetWidth}px`;
+        activeIndicatorElement.style.width = `${activeTriggerElement.offsetWidth}px`;
+    } else {
+        activeIndicatorElement.style.left = `${activeTriggerElement.offsetLeft}px`;
+        activeIndicatorElement.style.width = `${activeTriggerElement.offsetWidth}px`;
+    }
+};
+
+export const useTabTriggers = ({
+    activeTab,
+}: {
+    activeTab?: string;
+}): {
+    triggerListRef: RefObject<HTMLDivElement>;
+    activeIndicatorRef: RefObject<HTMLSpanElement>;
+    triggers: TabTrigger[];
+    triggersOutOfView: TabTrigger[];
+    addTrigger: (trigger: TabTrigger) => void;
+} => {
+    const triggerListRef = useRef<HTMLDivElement>(null);
+    const activeIndicatorRef = useRef<HTMLSpanElement>(null);
+    const [triggers, setTriggers] = useState<TabTrigger[]>([]);
+    const [triggersOutOfView, setTriggersOutOfView] = useState<TabTrigger[]>([]);
+
+    // move the active indicator and scroll to the correct position when the tab changes
+    useEffect(() => {
+        const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]');
+        if (activeTriggerElement) {
+            moveActiveIndicator(triggerListRef, activeIndicatorRef);
+            activeTriggerElement.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'end' });
+        }
+    }, [activeTab, activeIndicatorRef]);
+
+    useLayoutEffect(() => {
+        const activeTriggerElement = triggerListRef.current?.querySelector('[data-state="active"]') as HTMLElement;
+        const triggerListElement = triggerListRef.current as HTMLDivElement;
+
+        // move the active indicator to the initial active tab
+        if (activeTriggerElement) {
+            moveActiveIndicator(triggerListRef, activeIndicatorRef);
+        }
+
+        // calculate the overflowing elements when item starts overflowing the parent
+        const intersectionObserver = new IntersectionObserver(
+            () => {
+                setTriggersOutOfView(getOverflowingTriggers(triggers, triggerListElement));
+                moveActiveIndicator(triggerListRef, activeIndicatorRef);
+            },
+            {
+                root: triggerListElement,
+                threshold: 1,
+            },
+        );
+        for (const trigger of triggers) {
+            if (trigger.ref?.current) {
+                intersectionObserver.observe(trigger.ref.current);
```

Comment:
One interaction observer per trigger seems not so great, can't you have it on the parent instead?

Reply:
there is actually only one `intersectionObserver` 

the `observe()` function adds targets to it and is not creating new observers